src/control/lib/hardware/hwloc/provider.go

```diff
@@ -130,58 +134,123 @@ func (p *Provider) getNUMANodes(topo *topology) (map[uint]*hardware.NUMANode, er
 		}
 		prevNode = numaObj
 
-		nodeStr := numaObj.nodeSet().String()
-
 		id := numaObj.osIndex()
 		newNode := &hardware.NUMANode{
-			ID:       id,
-			NumCores: coresByNode[nodeStr],
-		}
-
-		if devs, exists := pciDevsByNode[id]; exists {
-			newNode.Devices = devs
-		} else {
-			newNode.Devices = make(hardware.PCIDevices)
+			ID:  id,
+			Set: numaObj.nodeSet().String(),
 		}
 
 		nodes[id] = newNode
 	}
 
 	if len(nodes) == 0 {
 		// If hwloc didn't detect any NUMA nodes, create a virtual NUMA node 0
-		totalCores := uint(0)
-		for _, numCores := range coresByNode {
-			totalCores += numCores
-		}
 		nodes[0] = &hardware.NUMANode{
-			ID:       0,
-			NumCores: totalCores,
-			Devices:  pciDevsByNode[0],
+			ID:  0,
+			Set: synthNodeSet,
 		}
 	}
 
+	if err := p.getCoresPerNodeSet(topo, nodes); err != nil {
+		return nil, err
+	}
+	if err := p.getPCIBridgesPerNUMANode(topo, nodes); err != nil {
+		return nil, err
+	}
+	if err := p.getPCIDevsPerNUMANode(topo, nodes); err != nil {
+		return nil, err
+	}
+
 	return nodes, nil
 }
 
-func (p *Provider) getCoreCountsPerNodeSet(topo *topology) map[string]uint {
+func (p *Provider) getCoresPerNodeSet(topo *topology, nodes nodeMap) error {
+	if topo == nil || nodes == nil {
+		return errors.New("nil topology or nodes")
+	}
+
 	prevCore := (*object)(nil)
-	coresPerNode := make(map[string]uint)
 	for {
 		coreObj, err := topo.getNextObjByType(objTypeCore, prevCore)
 		if err != nil {
 			break
 		}
 
-		coresPerNode[coreObj.nodeSet().String()]++
+		nodeID := coreObj.nodeSet().String()
+		for _, node := range nodes {
+			if node.Set != nodeID {
+				continue
+			}
+			node.AddCore(hardware.CPUCore{
+				ID: coreObj.logicalIndex(),
+			})
+		}
 
 		prevCore = coreObj
 	}
 
-	return coresPerNode
+	return nil
 }
 
-func (p *Provider) getPCIDevsPerNUMANode(topo *topology) map[uint]hardware.PCIDevices {
-	pciPerNUMA := make(map[uint]hardware.PCIDevices)
+func (p *Provider) getPCIBridgesPerNUMANode(topo *topology, nodes nodeMap) error {
+	if topo == nil || nodes == nil {
+		return errors.New("nil topology or nodes")
+	}
+
+	var bus *hardware.PCIBus
+	prev := (*object)(nil)
+	for {
+		bridge, err := topo.getNextObjByType(objTypeBridge, prev)
+		if err != nil {
+			// end of the list
+			break
+		}
+		prev = bridge
+
+		bridgeType, err := bridge.bridgeType()
+		if err != nil {
+			return err
+		}
+
+		numaID := p.getDeviceNUMANodeID(bridge, topo)
+		switch bridgeType {
+		case bridgeTypeHost:
+			lo, hi, err := bridge.busRange()
+			if err != nil {
+				return err
+			}
+
+			bus = &hardware.PCIBus{
+				LowAddress:  *lo,
+				HighAddress: *hi,
+			}
+			for _, node := range nodes {
+				if node.ID != numaID {
+					continue
+				}
+				node.AddPCIBus(bus)
+			}
+		case bridgeTypePCI:
+			if bus == nil {
+				return errors.New("unexpected PCI bridge before host bridge")
+			}
+			/*addr, err := bridge.pciAddr()
+			if err != nil {
+				return err
+			}
+			bus.AddressRange.Add(addr)*/
```

Comment:
Not working yet? Or forgotten?

Reply:
Cruft, will remove. Was from before I understood how the bridges were arranged.